File: example_mobile_app/ios/ExampleDBApp/ExampleDBApp/DBHelper.m
//
//  DBHelper.m
//

#import "DBHelper.h"
#import "AppDelegate.h"

@implementation DBHelper


#pragma mark - Setup Database
// please override these 2 methods to get the sqlite database
+ (FMDatabase*) getReadableDatabase
{
    /* example implementation */
    AppDelegate* sharedDelegate = [AppDelegate sharedAppDelegate];
    FMDatabase* db = [FMDatabase databaseWithPath:sharedDelegate.databasePath];
    return db;
}

+ (FMDatabase*) getWritableDatabase
{
    /* example implementation */
    AppDelegate* sharedDelegate = [AppDelegate sharedAppDelegate];
    FMDatabase* db = [FMDatabase databaseWithPath:sharedDelegate.databasePath];
    return db;
}


#pragma mark - CRUD
// CRUD for Users

// insert/update
+ (BOOL) insertUser:(User*)entry
{
    if(entry == nil)
        return NO;
    
    FMDatabase* db = [self getWritableDatabase];
    [db open];
    
    BOOL success = NO;
    
    // insert
    success =  [db executeUpdate:@"INSERT INTO User "
                "( userId,name,handphone,email ) VALUES "
                "( ?,?,?,?  ) "
                ,
                [NSNumber numberWithInt: entry.userId], entry.name, [NSNumber numberWithInt: entry.handphone], entry.email
                
                ];
    
    
    [db close];
    
    return success;
}

+ (BOOL) updateUser:(User*)entry
{
    if(entry == nil)
        return NO;
    
    FMDatabase* db = [self getWritableDatabase];
    [db open];
    
    BOOL success = NO;
    
    // update
    success =  [db executeUpdate:@"UPDATE User SET "
                " name = ? ,handphone = ? ,email = ?  "
                "WHERE userId = ?"
                ,
                entry.name, [NSNumber numberWithInt: entry.handphone], entry.email, [NSNumber numberWithInt: entry.userId]
                
                ];
    
    
    [db close];
    
    return success;
    
}

// check primary key, insert if not exist, update if exist
+ (BOOL) saveUser:(User*)entry
{
    if(entry == nil )
        return NO;
    
    BOOL isExist = [self isUserExist:entry.userId];
    if(isExist)
    {
        return [self updateUser:entry];
    }
    else{
        return [self insertUser:entry];
    }
    
}

// deletion
+ (BOOL) deleteAllUsers
{
    FMDatabase* db = [self getWritableDatabase];
    [db open];
    
    BOOL success = [db executeUpdate:@"DELETE FROM User"];
    
    [db close];
    
    return success;
}

+ (BOOL) deleteUserByPK:(int) userId
{
    FMDatabase* db = [self getWritableDatabase];
    [db open];
    
    BOOL success = [db executeUpdate:@"DELETE FROM User WHERE userId = ?",  [NSNumber numberWithInt: userId] ];
    [db close];
    
    return success;
}

+ (NSMutableArray*) queryUsers: (NSString*)queryString withArgumentsInArray:(NSArray*)args
{
    NSMutableArray* arr = [[NSMutableArray alloc]init];
    FMDatabase* db = [self getReadableDatabase];
    [db open];
    
    FMResultSet *results = [db executeQuery:queryString withArgumentsInArray:args];
    
    while([results next])
    {
        User* entry = [[User alloc]init];
        
        entry.userId = [results intForColumn:@"userId"];
        entry.name = [results stringForColumn:@"name"];
        entry.handphone = [results intForColumn:@"handphone"];
        entry.email = [results stringForColumn:@"email"];
        
        
        [arr addObject:entry];
        
    }
    
    [results close];
    
    [db close];
    
    return arr;
}

// query
+ (NSMutableArray*) queryUsers: (NSString*)queryString
{
    return [self queryUsers:queryString withArgumentsInArray:nil];
}

+ (NSMutableArray*) findAllUsers
{
    return [self queryUsers:@"SELECT * FROM User"];
}

+ (User*) findUserByPK:(int)userId
{
    NSArray* args =[ [NSArray alloc]initWithObjects: [NSNumber numberWithInt: userId], nil];
    NSMutableArray* results = [self queryUsers:@"SELECT * FROM User WHERE userId = ?" withArgumentsInArray:args];
    if(results.count > 0)
        return [results objectAtIndex:0];
    else
        return nil;
}

+ (BOOL) isUserExist:(int)userId
{
    User* entry = [self findUserByPK:userId];
    if(entry!=nil)
        return YES;
    else
        return NO;
}


@end

File: example_mobile_app/android/src/com/example/helper/DBHelper.java
package com.example.helper;  // TODO: PLEASE UPDATE THE PACKAGE NAME and import
import com.example.model.User;

import java.util.ArrayList;
import java.util.List;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.readystatesoftware.sqliteasset.SQLiteAssetHelper;  // wrapper for SQLiteAssetHelper

public class DBHelper  extends SQLiteAssetHelper{

	// the zip file must be testdb.zip
	// inside the zip file there must be a testdb-v1.db or testdb-v2.db
	// sqlitehelper dependencies. Please read documentation
	// the zip file must be under assets > databases folder
	private static final String DATABASE_NAME = "testdb";  // TODO: update the database name to be your database
	private static final int DATABASE_VERSION = 1;
	
	private static final String USER_TABLE_NAME = "User" ; // table name
	
	// column names 
    private static final String USERID = "userId" ; 
    private static final String NAME = "name" ; 
    private static final String HANDPHONE = "handphone" ; 
    private static final String EMAIL = "email" ; 

	
	
	private static final String[] USER_TABLE_COLUMNS = {  USERID ,NAME ,HANDPHONE ,EMAIL    } ;
	
	public DBHelper(Context context)
	{
		super(context, DATABASE_NAME, null, DATABASE_VERSION);
	}
	
	// CRUD for User

	public boolean insertUser(User entry)
	{
		if(entry!=null){
			SQLiteDatabase db = this.getWritableDatabase();
			ContentValues values = new ContentValues();
			
    values.put(USERID, entry.getUserId() ) ; 
    values.put(NAME, entry.getName() ) ; 
    values.put(HANDPHONE, entry.getHandphone() ) ; 
    values.put(EMAIL, entry.getEmail() ) ; 

			
			return (db.insert(USER_TABLE_NAME, null, values) != -1);	
		}
		
		
		return false;
	}
	
	public boolean updateUser(User entry)
	{
		if(entry == null)
			return false;
		
		SQLiteDatabase db = this.getWritableDatabase();
		ContentValues values = new ContentValues();
		
    values.put(USERID, entry.getUserId() ) ; 
    values.put(NAME, entry.getName() ) ; 
    values.put(HANDPHONE, entry.getHandphone() ) ; 
    values.put(EMAIL, entry.getEmail() ) ; 

		
		db.update(USER_TABLE_NAME, values, "userId = ?", 
									new String[]{ String.valueOf(  entry.getUserId() ) } );
		
		return true;
	}
	
	// check primary key, insert if not exist, update if exist
	public boolean saveUser(User entry)
	{
		if(entry == null)
			return false;
		
		boolean isExist = this.isUserExist(entry.getUserId() );
		
		if(isExist)
			return updateUser(entry);
		else
			return insertUser(entry);
		
	}
	
	// deletion
	public int deleteAllUsers()
	{
		SQLiteDatabase db = this.getWritableDatabase();
		int noOfRowsAffected = db.delete(USER_TABLE_NAME, null, null);
		return noOfRowsAffected;
	}
	
	public int deleteUserByPK(int userId)
	{
		SQLiteDatabase db = this.getWritableDatabase();
		int noOfRowsAffected = db.delete(USER_TABLE_NAME, "userId = ?", 
											new String[]{ String.valueOf(  userId ) });
		return noOfRowsAffected;
	}
	
	
	private List<User> processUserCursor(Cursor c)
	{
		List<User> results = new ArrayList<User>();
		
		int numRows = c.getCount();
		c.moveToFirst();
			
		for (int i = 0; i < numRows; i++) {
			
			User entry = new User();
			
			
int userId = c.getInt(0); 
    entry.setUserId(userId); 
String name = c.getString(1); 
    entry.setName(name); 
int handphone = c.getInt(2); 
    entry.setHandphone(handphone); 
String email = c.getString(3); 
    entry.setEmail(email); 

			
			results.add(entry);
			
			c.moveToNext();
		}
		
		c.close();
		
		return results;
	}
	
	public List<User> findAllUsers()
	{
		SQLiteDatabase db = getReadableDatabase();
		
		Cursor c = db.query(USER_TABLE_NAME, 
							USER_TABLE_COLUMNS, 
							null, null, // where arguments 
							null, null, null // groupby, having, orderby
				); 
		
		return processUserCursor(c);

	}
	

	public User findUserByPK(int userId )
	{
		
		SQLiteDatabase db = getReadableDatabase();
		
		Cursor c = db.query(USER_TABLE_NAME, 
							USER_TABLE_COLUMNS, 
							"userId = ?", new String[]{ String.valueOf(  userId ) }, // where arguments
							null, null, null // groupby, having, orderby
				); 
		
		List<User> results = processUserCursor(c);
		if(results == null || results.size() == 0)
			return null;
		else
			return results.get(0);
	}
	

	public boolean isUserExist( int userId )
	{
		User entry = findUserByPK(userId);
		return (entry!=null);
	}
	
	


		
}